src/log-store/src/kafka/log_store.rs

```diff
@@ -101,24 +101,30 @@ impl LogStore for KafkaLogStore {
         }
 
         // Builds a record from entries belong to a region and produces them to kafka server.
-        let region_ids = producers.keys().cloned().collect::<Vec<_>>();
-
-        let tasks = producers
-            .into_values()
-            .map(|producer| producer.produce(&self.client_manager))
-            .collect::<Vec<_>>();
+        let (region_ids, tasks): (Vec<_>, Vec<_>) = producers
+            .into_iter()
+            .map(|(id, producer)| (id, producer.produce(&self.client_manager)))
+            .unzip();
         // Each produce operation returns a kafka offset of the produced record.
         // The offsets are then converted to entry ids.
         let entry_ids = futures::future::try_join_all(tasks)
             .await?
             .into_iter()
             .map(TryInto::try_into)
             .collect::<Result<Vec<_>>>()?;
-        debug!("The entries are appended at offsets {:?}", entry_ids);
+        let last_entry_ids = region_ids
+            .into_iter()
+            .zip(entry_ids)
+            .collect::<HashMap<_, _>>();
```

Comment:
This would be cleaner:
```suggestion
        let last_entry_ids =
            futures::future::try_join_all(producers.into_iter().map(|(region_id, producer)| {
                Box::pin(async move {
                    let entry_id = producer.produce(&self.client_manager).await?;
                    Ok((region_id, entry_id.0 as u64))
                })
            }))
            .await?
            .into_iter()
            .collect::<HashMap<_, _>>();
```

Reply:
Does this work without `Box::pin`?